pandas/io/html.py

```diff
@@ -358,47 +336,115 @@ def _build_doc(self):
 
         Returns
         -------
-        obj : tree-like
+        obj : the DOM from which to parse the table element.
         """
         raise com.AbstractMethodError(self)
 
-    def _build_table(self, table):
-        header = self._parse_raw_thead(table)
-        body = self._parse_raw_tbody(table)
-        footer = self._parse_raw_tfoot(table)
-        return header, body, footer
+    def _parse_thead_tbody_tfoot(self, table_html):
+        """Given a table, return parsed header, body, and foot.
+           Header and body are lists-of-lists. Top level list is a list of
+           rows. Each row is a list of parsed elements.
 
-    def _parse_raw_thead(self, table):
-        thead = self._parse_thead(table)
-        res = []
-        if thead:
-            trs = self._parse_tr(thead[0])
-            for tr in trs:
-                cols = lmap(self._text_getter, self._parse_td(tr))
-                if any(col != '' for col in cols):
-                    res.append(cols)
-        return res
+           Logic: Use <thead>, <tbody>, <tfoot> elements to identify
+                  header, body, and footer, otherwise:
+                  - Put all rows into body
+                  - Move rows from top of body to header only if
+                    all elements inside row are <th>
+                  - Move rows from bottom of body to footer only if
+                    all elements inside row are <th>
 
-    def _parse_raw_tfoot(self, table):
-        tfoot = self._parse_tfoot(table)
-        res = []
-        if tfoot:
-            res = lmap(self._text_getter, self._parse_td(tfoot[0]))
-        return np.atleast_1d(
-            np.array(res).squeeze()) if res and len(res) == 1 else res
+        Parameters
+        ----------
+        table_html : a single HTML table element.
 
-    def _parse_raw_tbody(self, table):
-        tbodies = self._parse_tbody(table)
+        Returns
+        -------
+        tuple of (header, body, footer)
+        header : list of rows, each of which is a list of parsed
+                 header elements
+        body : list of rows, each of which is a list of parsed body elements
+        footer : list of rows, each of which is a list of parsed
+                 footer elements
+        """
 
-        raw_data = []
+        header_rows = self._parse_thead_tr(table_html)
+        body_rows = self._parse_tbody_tr(table_html)
+        footer_rows = self._parse_tfoot_tr(table_html)
 
-        if tbodies:
-            for tbody in tbodies:
-                raw_data.extend(self._parse_tr(tbody))
-        else:
-            raw_data.extend(self._parse_tr(table))
+        if not header_rows:
+            # The table has no <thead>. Treat first all-<th> rows as headers.
+            while body_rows and all(self._equals_tag(t, 'th') for t in
```

Comment:
Could be overlooking something but why are you using `while` here - are you not just testing the truthiness of `body_rows`? If so, `if` would be preferable

Reply:
`body_rows` changes length each iteration. I just made the conditional shorter to make that clearer.

Reply:
Ah OK that makes more sense. Is that a safe operation though? Can't say I've been personally bitten by this but SO seems to advise against modifying the container that you loop over:

https://stackoverflow.com/questions/1637807/modifying-list-while-iterating

Reply:
Ah OK that makes more sense. Is that a safe operation though? Can't say I've been personally bitten by this but SO seems to advise against modifying the container that you loop over:

https://stackoverflow.com/questions/1637807/modifying-list-while-iterating

Reply:
I'd argue that this loop doesn't iterate over the _list_. It just iterates over a condition.